ss58-js/mod.ts

```diff
@@ -0,0 +1,89 @@
+import { blake2b } from "../deps/hashes.ts";
+import * as base58 from "../deps/std/encoding/base58.ts";
+
+import { hex } from "../util/mod.ts";
+
+import {
+  ALLOWED_ADDRESS_LENGTHS,
+  ALLOWED_NETWORK_PREFIXES,
+  ALLOWED_PUBLIC_KEY_LENGTHS,
+  CHECKSUM_LENGTH,
+  SS58PRE,
+} from "./constants.ts";
+
+export const encode = (prefix: number, pubKey: string): string => {
+  const pubKeyBytes = hex.decode(pubKey);
+
+  if (!ALLOWED_PUBLIC_KEY_LENGTHS.includes(pubKeyBytes.length)) {
+    throw new Error("Invalid public key length");
+  }
+
+  if (!ALLOWED_NETWORK_PREFIXES.includes(prefix)) {
+    throw new Error("Invalid network prefix");
+  }
+
+  const prefixBytes = prefix < 64
+    ? Uint8Array.of(prefix)
+    : Uint8Array.of(
+      ((prefix & 0b0000_0000_1111_1100) >> 2) | 0b0100_0000,
+      (prefix >> 8) | ((prefix & 0b0000_0000_0000_0011) << 6),
+    );
+
+  const hasher = blake2b.create({
+    dkLen: 512 / 8,
+  });
+
+  hasher.update(SS58PRE);
+  hasher.update(prefixBytes);
+  hasher.update(pubKeyBytes);
+
+  const digest = hasher.digest();
+  const checksum = new Uint8Array(digest.buffer, 0, CHECKSUM_LENGTH);
+
+  const addrBytes = new Uint8Array(prefixBytes.length + pubKeyBytes.length + CHECKSUM_LENGTH);
+
+  addrBytes.set(prefixBytes, 0);
+  addrBytes.set(pubKeyBytes, prefixBytes.length);
+  addrBytes.set(checksum, prefixBytes.length + pubKeyBytes.length);
+
+  return base58.encode(addrBytes);
+};
+
+export const decode = (address: string): [number, string] => {
```

Comment:
Like with `encode`, the raw logic should be decoupled from the string-encoding, so that raw Uint8Arrays can be passed around internally. Additionally, the tuple type should have labels for documentation purposes.
```suggestion
export const decode = (address: Uint8Array): [prefix: number, pubKey: Uint8Array] => {
```